File: trees/226_invert_binary_tree.py
# Approach 1: Breath First Search
# Time complexity: O(n)
# Space complexity: O(n) for queue.

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root: 
            return root

        queue = deque([root])
        while queue:
            node = queue.popleft()

            node.left, node.right = node.right, node.left
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        return root

# Approach 2: Depth First Search
# Time complexity: O(n)
# Space complexity: O(n) for call stack.

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root
        
        self.invertTree(root.left)
        self.invertTree(root.right)

        root.left, root.right = root.right, root.left

        return root

# Approach 3: Iterative DFS
# Time complexity: O(n)
# Space complexity: O(n) for stack.

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root

        stack = [root]

        while stack:
            node = stack.pop()
            node.left, node.right = node.right, node.left
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
        
        return root

File: trees/226_invert_binary_tree/226_invert_binary_tree.java
// Approach 1: Breath First Search
// Time complexity: O(n)
// Space complexity: O(n) for queue.

class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null;
        }
      
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
      
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            TreeNode temp = node.left;
            node.left = node.right;
            node.right = temp;
            if (node.left != null) {
                queue.offer(node.left);
            }
            if (node.right != null) {
                queue.offer(node.right);
            }
        }
      
        return root;
    }
}

// Approach 2: Depth First Search
// Time complexity: O(n)
// Space complexity: O(n) for call stack.

class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        
        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;

        invertTree(root.left);
        invertTree(root.right);

        return root;

    }
}

// Approach 3: Iterative DFS
// Time complexity: O(n)
// Space complexity: O(n) for stack.

class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null;
        }

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            TreeNode temp = node.left;
            node.left = node.right;
            node.right = temp;
            if (node.left != null) {
                stack.push(node.left);
            }
            if (node.right != null) {
                stack.push(node.right);
            }
        }
        return root;

    }
}
